ragna/deploy/_ui/modal_configuration.py

```diff
@@ -115,12 +124,34 @@ def __init__(self, api_wrapper, **params):
 
         self.got_timezone = False
 
+        self.corpus_or_upload_radiobutton = pn.widgets.RadioBoxGroup.from_param(
+            self.param.corpus_or_upload,
+            inline=True,
+            title="",
+            css_classes=["modal_configuration_radio_button_group"],
+        )
+
+        self.metadata_filters_builder = MetadataFiltersBuilder()
+
     def did_click_on_start_chat_button(self, event):
-        if not self.document_uploader.can_proceed_to_upload():
-            self.change_upload_files_label("missing_file")
-        else:
+        if self.corpus_or_upload == USE_UPLOAD_LABEL:
+            if not self.document_uploader.can_proceed_to_upload():
+                self.change_upload_files_label("missing_file")
+            else:
+                self.start_chat_button.disabled = True
+                self.document_uploader.perform_upload(event, self.did_finish_upload)
+
+        elif self.corpus_or_upload == USE_CORPUS_LABEL:
+            if not self.metadata_filters_builder.validate():
+                # The Metadata Filters Builder updates its UI when calling validate()
+                return
+
             self.start_chat_button.disabled = True
-            self.document_uploader.perform_upload(event, self.did_finish_upload)
+            asyncio.ensure_future(
+                self.did_finish_upload(
+                    self.metadata_filters_builder.get_metadata_filters()
+                )
+            )
```

Comment:
Why not make this method async and just `await` the result?

Reply:
The function is a button callback. I have had issues in the past with making such callbacks async. But that seems to work now, so let's do it.